### PR TITLE
INTYGFV-14571: Logic to update certain metadata in an existing termination.

### PR DESCRIPTION
**Pull request by mww-sopra-steria**

### CODE REVIEW

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/application/service/TerminationService.java [hunk 1]
```diff
@@ -50,6 +58,32 @@ public TerminationDTO create(CreateTerminationDTO createTerminationDTO) {
     return toDTO(createdTermination);
   }
 
+  public TerminationDTO update(UUID terminationId, TerminationDTO terminationDTO) {
+    if (!terminationId.equals(terminationDTO.terminationId())) {
```
Comment (IamFriberg) [inline customer-terminate-service/app/src/main/java/se/inera/intyg/cts/application/service/TerminationService.java:62]: Feel a bit redundant to pass the same parameter 2 times to this method. Unless we can update all the fields in TerminationDTO we should not use it as a parameter. (Now we only update/use 4 of 10 fields/arguments provided.)

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/persistence/JpaCertificateRepository.java [hunk 2]
```diff
@@ -51,6 +51,23 @@ public List<Certificate> get(Termination termination) {
         .collect(Collectors.toList());
   }
 
+  @Override
+  public void remove(Termination termination) {
+    final var terminationEntity = terminationEntityRepository
+        .findByTerminationId(termination.terminationId().id());
+
+    if (terminationEntity.isEmpty()) {
+      return;
```
Comment (IamFriberg) [inline customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/persistence/JpaCertificateRepository.java:60]: Maybe we should throw a NotFoundException if we expect to find a termination to remove?

**Review by IamFriberg — CHANGES_REQUESTED** [review]
Nice! Only have som smaller comments around the amount of field we pass to the update.

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/persistence/JpaCertificateRepository.java (hunk 2 above)
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/persistence/JpaCertificateRepository.java:60]: You have a good point. It is better to do as you suggest as it is correct that we do expect a termination to exist. 

Will fix!

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/persistence/JpaCertificateTextRepository.java [hunk 3]
```diff
@@ -52,4 +52,24 @@ public List<CertificateText> get(Termination termination) {
         .map(CertificateTextEntityMapper::toDomain)
         .collect(Collectors.toList());
   }
+
+  @Override
+  public void remove(Termination termination) {
+    final var terminationEntity = getByTerminationId(termination);
+
+    if (terminationEntity.isEmpty()) {
+      return;
```
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/persistence/JpaCertificateTextRepository.java:61]: You have a good point. It is better to do as you suggest as it is correct that we do expect a termination to exist. 

Will fix!

File: customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/UpdateTerminationImpl.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+  @Override
+  public Termination update(Termination termination, HSAId hsaId, PersonId personId,
+      EmailAddress emailAddress, PhoneNumber phoneNumber) {
+
+    termination.update(hsaId, personId, emailAddress, phoneNumber);
+
+    final var storeTermination = terminationRepository.store(termination);
+    if (storeTermination.status() == TerminationStatus.CREATED) {
+      certificateRepository.remove(storeTermination);
```
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/UpdateTerminationImpl.java:35]: We don't remove the termination, we remove any certificates and certificate texts that are related to the termination. 

It is an important behaviour for the service and there are unit tests that verifies it

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/application/service/TerminationService.java (hunk 1 above)
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/application/service/TerminationService.java:62]: I didn't want to put that logic in the controller, instead this app service verifies the data received through the api and then calls the domain service with the fields to update. If in tomorrow we allow another field to be updated, we don't have to change the api.

Comment (IamFriberg) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/application/service/TerminationService.java:62]: With this solution we allow the caller to provide arguments that are not used but can be provided. This make the API unsafe as we can not trust the service to updated all fields we provide. Another solution would be to use a updateTerminationDTO what only contains the 3 fields that will be used and give us the option to add optional fields in the future that will not affect the API.

**Review by IamFriberg — CHANGES_REQUESTED** [review]

File: customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/UpdateTerminationImpl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (IamFriberg) [reply customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/UpdateTerminationImpl.java:35]: Should we add some description to why we do this? My picture is that this is a part of a chain of events that does not realy make sense unless you know the whole chain and where it fits. :)

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/persistence/JpaCertificateRepository.java (hunk 2 above)
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/persistence/JpaCertificateRepository.java]: Fixed!

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/persistence/JpaCertificateTextRepository.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/persistence/JpaCertificateTextRepository.java]: Fixed!

File: customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/UpdateTerminationImpl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/UpdateTerminationImpl.java:35]: You have a good point. I will some javadoc that points out this important behaviour.

[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/UpdateTerminationImpl.java:39]: Fixed!

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/application/api/TerminationController.java [hunk 5]
```diff
@@ -37,6 +37,16 @@ TerminationDTO create(@RequestBody CreateTerminationDTO request) {
     }
   }
 
+  @PostMapping("/{terminationId}")
+  TerminationDTO update(@PathVariable UUID terminationId,
```
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/application/api/TerminationController.java:41]: It makes sense, and I think is expected, that the apis url specifies which termination to update. That it also exists in the TerminationDTO is a bit redundant, but is validated in the TerminationService that there is no missmatch.

Comment (IamFriberg) [inline customer-terminate-service/app/src/main/java/se/inera/intyg/cts/application/api/TerminationController.java:41]: With this solution we allow the caller to provide arguments that are not used but can be provided. This make the API unsafe as we can not trust the service to updated all fields we provide. Another solution would be to use a updateTerminationDTO what only contains the 3 fields that will be used and give us the option to add optional fields in the future that will not affect the API.

**Review by IamFriberg — CHANGES_REQUESTED** [review]

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/application/api/TerminationController.java (hunk 5 above)
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/application/api/TerminationController.java:41]: OK! I do get your point and I agree. Will make the change and communicate with @mikeyOne to make necessary changes in Intygsadmin.

Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/application/api/TerminationController.java:42]: Fixed!

**Review by IamFriberg — APPROVED** [review]
Nice!

**Pull request merged**